pkg/hydrator/nexpose.go

```diff
@@ -2,38 +2,324 @@ package hydrator
+}
+
+// NexposeClient implements the interfaces to fetch vulnerabilities and solutions from nexpose
+type NexposeClient struct {
+	Client   *http.Client
+	Host     *url.URL
+	Username string
+	Password string
+	PageSize int
+}
+
+// FetchVulnerabilitySolutions fetches the solutions to a particular vulnerability
+func (n *NexposeClient) FetchVulnerabilitySolutions(ctx context.Context, vulnID string) ([]string, error) {
+	body, err := n.makeNexposeRequest(nil, "api", "3", "vulnerabilities", vulnID, "solutions")
+	if err != nil {
+		return nil, err
+	}
+	var solutions vulnerabilitySolutions
+	err = json.Unmarshal(body, &solutions)
+	if err != nil {
+		return nil, err
+	}
+	return solutions.Resources, nil
+}
+
+// FetchSolution fetches details about a particular solution
+func (n *NexposeClient) FetchSolution(ctx context.Context, solutionID string) (string, error) {
+	body, err := n.makeNexposeRequest(nil, "api", "3", "solutions", solutionID)
+	if err != nil {
+		return "", err
+	}
+	var solutionDetails solution
+	err = json.Unmarshal(body, &solutionDetails)
+	if err != nil {
+		return "", err
+	}
+	return solutionDetails.Steps.Text, nil
+}
+
+// FetchVulnerabilityDetails fetches details about a particular vulnerability
+func (n *NexposeClient) FetchVulnerabilityDetails(ctx context.Context, vulnID string) (NexposeVulnerability, error) {
+	body, err := n.makeNexposeRequest(nil, "api", "3", "vulnerabilities", vulnID)
+	if err != nil {
+		return NexposeVulnerability{}, err
+	}
+	var vulnDetails vulnerabilityDetails
+	err = json.Unmarshal(body, &vulnDetails)
+	if err != nil {
+		return NexposeVulnerability{}, err
+	}
+	return NexposeVulnerability{
+		Title:          vulnDetails.Title,
+		Description:    vulnDetails.Description.Text,
+		CvssV2Score:    vulnDetails.CVSS.V2.Score,
+		CvssV2Severity: cvssV2Severity(vulnDetails.CVSS.V2.Score),
+	}, nil
+}
+
+// FetchAssetVulnerabilities fetches the list of vulnerabilities for an asset
+func (n *NexposeClient) FetchAssetVulnerabilities(ctx context.Context, assetID int64) ([]NexposeAssetVulnerability, error) {
+	body, err := n.makeNexposeRequest(
+		map[string]string{pageQueryParam: "0", sizeQueryParam: strconv.Itoa(n.PageSize)},
+		"api", "3", "assets", strconv.FormatInt(assetID, 10), "vulnerabilities",
+	)
+	if err != nil {
+		return nil, err
+	}
+	var assetVulns assetVulnerabilities
+	err = json.Unmarshal(body, &assetVulns)
+	if err != nil {
+		return nil, err
+	}
+
+	pages := int(assetVulns.Page.TotalPages)
+	totalResources := int(assetVulns.Page.TotalResources)
+	assetVulnsChannel := make(chan NexposeAssetVulnerability, totalResources)
+	errorChan := make(chan error, pages)
+	for _, resource := range assetVulns.Resources {
+		assetVulnsChannel <- assetVulnToNexposeAssetVuln(resource)
+	}
+
+	for curPage := 1; curPage < pages; curPage = curPage + 1 {
+		go n.makePagedAssetVulnerabilitiesRequest(assetID, curPage, assetVulnsChannel, errorChan)
```

Comment:
task: While it isn't always a bad thing, passing channels into a function that writes its results to those channels should be considered a code smell. In this case you're actually doing a fan-out operation but in a slightly different way than the pattern.

I can be convinced that we don't need to use a stack of interfaces for this specific case but I'd suggest that we at least convert each function currently taking a success+error channel pair to a function that returns a value and an error. The channel management should be the responsibility of the code performing the fan-out. For example, this code would end up looking like this:

```golang
pages := int(assetVulns.Page.TotalPages)
totalResources := int(assetVulns.Page.TotalResources)
assetVulnsResults := make([] NexposeAssetVulnerability, 0, totalResources)
assetVulnsChannel := make(chan []NexposeAssetVulnerability, totalResources)
errorChan := make(chan error, pages)
for _, resource := range assetVulns.Resources {
    assetVulnsResults = append(assetVulnsResults, assetVulnToNexposeAssetVuln(resource))
}
for curPage := 1; curPage < pages; curPage = curPage + 1 {
    go func(curPage int) {
        vulns, err := n.makePagedAssetVulnerabilitiesRequest(assetID, curPage)
        if err != nil {
            errorChan <- err
            return
        }
        assetVulnsChannel <- vulns
    }(curPage)
}
// ... the rest is pretty much the same
```
There will be cases where we decide to pass channels around but they should be relatively rare or infrequent. In the vast majority of cases, our functions should do a task and optionally return a value. All concurrency related matters should be left to the code that introduces that concurrency rather than trying to negotiate concurrency from the bottom of the stack up.